proxy/conn.go

```diff
@@ -0,0 +1,114 @@
+package auth
+
+import (
+	"encoding/binary"
+	"net"
+
+	. "github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/packet"
+	"github.com/siddontang/go/hack"
+)
+
+type Conn struct {
+	clientConn *packet.Conn
+	serverConn *packet.Conn
+}
+
+func NewProxyConn(client, server net.Conn) *Conn {
+	c := packet.NewConn(client)
+	s := packet.NewConn(server)
+	return &Conn{clientConn: c, serverConn: s}
+}
+
+func (c *Conn) Handshake() error {
+	for {
+		mysqlPacket, err := c.forwardServerPacket()
+		if err != nil {
+			return err
+		}
+
+		if mysqlPacket[0] == OK_HEADER {
+			return nil
+		}
+		if mysqlPacket[0] == ERR_HEADER {
+			return c.handleErrorPacket(mysqlPacket)
+		}
+
+		if mysqlPacket[0] == MORE_DATE_HEADER {
+			switch mysqlPacket[1] {
+			case CACHE_SHA2_FAST_AUTH:
+				mysqlPacket, err = c.forwardServerPacket()
+				if err != nil {
+					return err
+				}
+			case CACHE_SHA2_FULL_AUTH:
+				err = c.forwardClientPacket()
+				if err != nil {
+					return err
+				}
+
+				mysqlPacket, err = c.forwardServerPacket()
+				if err != nil {
+					return err
+				}
+			}
+		}
+		if mysqlPacket[0] == OK_HEADER {
+			return nil
+		}
+
+		if mysqlPacket[0] == ERR_HEADER {
+			return c.handleErrorPacket(mysqlPacket)
+		}
+
+		err = c.forwardClientPacket()
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *Conn) forwardClientPacket() error {
+	clientPacket, err := c.clientConn.ReadPacket()
+	if err != nil {
+		return err
+	}
+
+	err = c.serverConn.WritePacket(append([]byte{0, 0, 0, 0}, clientPacket...))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Conn) forwardServerPacket() ([]byte, error) {
+	mysqlPacket, err := c.serverConn.ReadPacket()
+	if err != nil {
+		return nil, err
+	}
+	err = c.clientConn.WritePacket(append([]byte{0, 0, 0, 0}, mysqlPacket...))
+	if err != nil {
+		return nil, err
+	}
+	return mysqlPacket, nil
+}
+
+func (c *Conn) handleErrorPacket(data []byte) error {
```

Comment:
can we re-use the method in `client`, or is it too difficult?

Reply:
I wanted to, but there is a capability check in there since they actually unmarshal the packet and stuff so it makes it difficult as we dont have any of the things on that struct.

Reply:
something like

```go

func handleErrorPacket(data []byte) error {
	e := new(MyError)

	var pos = 1

	e.Code = binary.LittleEndian.Uint16(data[pos:])
	pos += 2

	if c.capability&CLIENT_PROTOCOL_41 > 0 {
		//skip '#'
		pos++
		e.State = hack.String(data[pos : pos+5])
		pos += 5
	}

	e.Message = hack.String(data[pos:])

	return e
}

func (*c Conn handleErrorPacket(data []byte) error {
   return handleErrorPacket(data)
}
```

don't work?

Reply:
oh right, `c.capability`